src/models/PartialFormSubmission.php

```diff
@@ -93,123 +145,154 @@ public function getCMSFields()
+        );
     }
 
     /**
-     * @param Member $member
-     * @param array $context
-     * @return bool
+     * Generate a key based on the share token salt
+     *
+     * @param string $token
+     * @return mixed
      */
-    public function canCreate($member = null, $context = [])
+    public function generateKey($token)
     {
-        return false;
+        return hash_pbkdf2('sha256', $token, $this->TokenSalt, 1000, 16);
     }
 
     /**
-     * @param Member
-     *
-     * @return boolean|string
+     * Generate the partial tokens
+     * If the submission is password protected, generate a password.
+     * @throws Exception
      */
-    public function canView($member = null)
+    public function onBeforeWrite()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canView($member);
-        }
-
-        return parent::canView($member);
+        parent::onBeforeWrite();
+        $this->getPartialToken();
+        $this->generatePassword();
     }
 
     /**
-     * @param Member
+     * Get the unique token for the share link
      *
-     * @return boolean|string
+     * @return bool|string|null
+     * @throws Exception
      */
-    public function canEdit($member = null)
+    protected function getPartialToken()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canEdit($member);
+        if (!$this->TokenSalt) {
+            $this->TokenSalt = $this->generateToken();
+            $this->Token = $this->generateToken();
         }
 
-        return parent::canEdit($member);
+        return $this->Token;
     }
 
     /**
-     * @param Member
+     * Generate a new token
      *
-     * @return boolean|string
+     * @return bool|string
+     * @throws Exception
      */
-    public function canDelete($member = null)
+    protected function generateToken()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canDelete($member);
-        }
+        $generator = new RandomGenerator();
 
-        return parent::canDelete($member);
+        return substr($generator->randomToken('sha256'), 0, 16);
     }
 
     /**
-     * Get the share link of the form
-     *
-     * @return string
-     * @throws \Exception
+     * @return void
      */
-    public function getPartialLink()
+    protected function generatePassword()
     {
-        if (!$this->isInDB()) {
-            return '(none)';
+        if (!$this->Password) {
+            $chars = range('A', 'Z');
+            $chars = array_merge($chars, range('a', 'z'));
+            $chars = array_merge($chars, range(0, 9));
+            $chars = array_merge($chars, self::$special_characters);
+            shuffle($chars);
+            $pwd = implode(array_slice($chars, 0, 10));
+            $this->Password = hash_pbkdf2('SHA256', $pwd, $this->TokenSalt, 1000);
+            Controller::curr()->getRequest()->getSession()->set('FormPassword', $pwd);
```

Comment:
Hmm.. shouldn't be passwords set in the session in the "page 0" anyway? Also, with one-way hashing how are we going to show that in the "page 0" form?

Reply:
_if_ the page-0 is used, then yes, it should be shown there.

The password is set in the session, exactly to be able to retrieve it on Page-0 or when going to the "save and continue later" (which is basically the same thing